text/0105-load-based-replica-read.md

```diff
@@ -0,0 +1,98 @@
+
+The client decides the maximum waiting duration. If this field is set, TiKV can return `ServerIsBusy` early if the estimated waiting duration exceeds the threshold. This threshold is also configurable by the user.
+
+```protobuf
+message Context {
+    ...
+
+    uint32 busy_threshold_ms = 26;
+}
+```
+
+If TiKV returns `ServerIsBusy` based on the current load, the estimated waiting duration and the current index on the leader will be also returned to the client.
+
+```protobuf
+message ServerIsBusy {
+    ...
+
+    uint32 estimated_wait_ms = 3;
+    uint64 applied_index = 4;
+}
+```
+
+Because we will retry in replica-read mode, we don't need the follower or learner to issue a read index RPC again after knowing the applied index.
+
+The estimated waiting duration may be useful for the client to select a better TiKV for the retried request.
+
+### Retry strategy
+
+_It is hard to find a perfect strategy. It is expected that the strategy changes after further experiemnts. So, this part only describes a possible solution._
+
+Without any other information, the client will select a random follower or learner to retry the request after the client receives a `ServerIsBusy`. The `busy_threshold_ms` in the retried request will be set to _2 times_ the one returned in `ServerIsBusy`. So, if the retried TiKV is _much_ busier than the leader, it will reject the request again. The factor is to balance the extra RPC number and the load difference between TiKVs.
+
+When the retried request is rejected again, the client will select a not selected replica to retry again. The `busy_threshold_ms` should still be calculated from the leader response. If all followers and learners reject the request because they are much busier than the leader, just unset `busy_threshold_ms` and let the leader execute the request without checking its load.
+
+#### Optimizing strategy with load info
+
+If the non-leader replicas are all busier than the leader, with the default strategy, we need `replica_num + 1` RPCs for each request. This is definitely unsatisfactory. Therefore, we should try to avoid useless attempts to busy nodes.
+
+We maintain the estimated waiting duration in the client. Every time the client receives a `ServerIsBusy` error with `estimated_wait_ms`, it updates the information for the store and records the current time.
+
+```go
+type Store sturct {
+    ...
+    
+    estimatedWait     time.Duration
+    waitTimeUpdatedAt time.Time
+}
+```
+
+To make use of as many resources as possible, the load we predict should not be larger than the current load. Otherwise, we may skip a node that is already free for executing requests and not get the best performance.
+
+We use `estimatedWait - (time.Now().Since(waitTimeUpdatedAt))` as the estimated waiting duration in the client. It's mostly certain that this estimated value is smaller than real because the TiKV accepts requests meanwhile and some queries don't finish in a single time slice.
```

Comment:
My initial thought is to let the client use the observed metrics like `cop_task_avg_wait_duration` in a recent time interval or something like that to decide which replica to choose next. This one looks simpler and could avoid retrying already busy replicas :thinking: